src/Components/Web.JS/src/Platform/WebAssemblyComponentAttacher.ts

```diff
@@ -0,0 +1,293 @@
+  type: string;
+  id: number;
+  typeName: string;
+  assembly: string;
+  parameterDefinitions: string | undefined;
+  parameterValues: string | undefined;
+}
+
+export class ComponentDescriptor {
+  private static globalId = 1;
+  public type: string;
+
+  public typeName: string;
+
+  public assembly: string;
+
+  public parameterDefinitions: string | undefined;
+
+  public parameterValues: string | undefined;
+
+  public id: number;
+
+  public start: Node;
+
+  public end?: Node;
+
+  public constructor(type: string, start: Node, end: Node | undefined, assembly: string, typeName: string, parameterDefinitions: string | undefined, parameterValues: string | undefined) {
+    this.id = ComponentDescriptor.globalId++;
+    this.type = type;
+    this.assembly = assembly;
+    this.typeName = typeName;
+    this.parameterDefinitions = parameterDefinitions;
+    this.parameterValues = parameterValues;
+    this.start = start;
+    this.end = end;
+  }
+
+  public toRecord(): ComponentMarker {
+    const result = { type: this.type, id: this.id, assembly: this.assembly, typeName: this.typeName, parameterDefinitions: this.parameterDefinitions, parameterValues: this.parameterValues };
+    return result;
+  }
+}
+
+export function discoverComponents(document: Document): ComponentDescriptor[] {
+  const componentComments = resolveComponentComments(document);
+  const discoveredComponents: ComponentDescriptor[] = [];
+  for (let i = 0; i < componentComments.length; i++) {
+    const componentComment = componentComments[i];
+    const entry = new ComponentDescriptor(
+      componentComment.type,
+      componentComment.start,
+      componentComment.end,
+      componentComment.assembly,
+      componentComment.typeName,
+      componentComment.parameterDefinitions,
+      componentComment.parameterValues,
+    );
+
+    discoveredComponents.push(entry);
+  }
+
+  return discoveredComponents.sort((a, b) => a.id - b.id);
+}
+
+interface ComponentComment {
+  type: 'server' | 'client';
+  assembly: string;
+  typeName: string;
+  parameterDefinitions?: string;
+  parameterValues?: string;
+  start: Node;
+  end?: Node;
+  prerenderId?: string;
+}
+
+function resolveComponentComments(node: Node): ComponentComment[] {
+  if (!node.hasChildNodes()) {
+    return [];
+  }
+
+  const result: ComponentComment[] = [];
+  const childNodeIterator = new ComponentCommentIterator(node.childNodes);
+  while (childNodeIterator.next() && childNodeIterator.currentElement) {
+    const componentComment = getComponentComment(childNodeIterator);
+    if (componentComment) {
+      result.push(componentComment);
+    } else {
+      const childResults = resolveComponentComments(childNodeIterator.currentElement);
+      for (let j = 0; j < childResults.length; j++) {
+        const childResult = childResults[j];
+        result.push(childResult);
+      }
+    }
+  }
+
+  return result;
+}
+
+const blazorCommentRegularExpression = /\W*Blazor:[^{]*(.*)$/;
```

Comment:
There wasn’t named capture groups support at the time this was written, lol.

I’ll address this in a follow-up cleanup PR if the build passes, I want to avoid triggering a separate build and we can probably also change it on the server code.

Reply:
done